clients/src/main/java/org/apache/kafka/clients/consumer/internals/ListOffsetsRequestManager.java

```diff
@@ -0,0 +1,359 @@
+        requireNonNull(time);
+        requireNonNull(apiVersions);
+        requireNonNull(logContext);
+
+        this.metadata = metadata;
+        this.metadata.addClusterUpdateListener(this);
+        this.isolationLevel = isolationLevel;
+        this.log = logContext.logger(getClass());
+        this.requestsToRetry = new HashSet<>();
+        this.requestsToSend = new ArrayList<>();
+        this.offsetFetcherUtils = new OffsetFetcherUtils(logContext, metadata, subscriptionState,
+                time, apiVersions);
+    }
+
+    /**
+     * Determine if a there are pending fetch offsets requests to be sent and build a
+     * {@link org.apache.kafka.clients.consumer.internals.NetworkClientDelegate.PollResult}
+     * containing it.
+     */
+    @Override
+    public NetworkClientDelegate.PollResult poll(final long currentTimeMs) {
+        if (requestsToSend.isEmpty()) {
+            return new NetworkClientDelegate.PollResult(Long.MAX_VALUE, Collections.emptyList());
+        }
+
+        NetworkClientDelegate.PollResult pollResult =
+                new NetworkClientDelegate.PollResult(Long.MAX_VALUE, new ArrayList<>(requestsToSend));
+        this.requestsToSend.clear();
+
+        return pollResult;
+    }
+
+    /**
+     * Retrieve offsets for the given partitions and timestamp.
+     *
+     * @param partitions        Partitions to get offsets for
+     * @param timestamp         Target time to look offsets for
+     * @param requireTimestamps True if this should fail with an UnsupportedVersionException if
+     *                          the broker does not support fetching precise timestamps for offsets
+     * @return Future containing the map of offsets retrieved for each partition. The future will
+     * complete when the responses for the requests are received and processed following a call
+     * to {@link #poll(long)}
+     */
+    public CompletableFuture<Map<TopicPartition, Long>> fetchOffsets(final Set<TopicPartition> partitions,
+                                                                     final long timestamp,
+                                                                     final boolean requireTimestamps) {
+        metadata.addTransientTopics(offsetFetcherUtils.topicsForPartitions(partitions));
+
+        Map<TopicPartition, Long> timestampsToSearch = partitions.stream()
+                .collect(Collectors.toMap(Function.identity(), tp -> timestamp));
+        ListOffsetsRequestState listOffsetsRequestState = new ListOffsetsRequestState(
+                timestampsToSearch,
+                requireTimestamps,
+                offsetFetcherUtils,
+                isolationLevel);
+        listOffsetsRequestState.globalResult.whenComplete((result, error) -> {
+            metadata.clearTransientTopics();
+            log.debug("Fetch offsets completed for partitions {} and timestamp {}. Result {}, " +
+                    "error", partitions, timestamp, result, error);
```

Comment:
I get your point, but the issue here is that Log4j treats any throwable passed as last argument differently. So, having `error: {}` would lead to a warning of "more placeholders than args", along with a weird log line having `"Fetch offsets completed for partitions [t1-1] and timestamp -2. Result null, error {}(org.apache.kafka.clients.consumer.internals.ListOffsetsRequestManager:138)"`

The way the log shows with the current way is ```
"Fetch offsets completed for partitions [t1-1] and timestamp -2. Result null, error (org.apache.kafka.clients.consumer.internals.ListOffsetsRequestManager:138)
org.apache.kafka.common.errors.TopicAuthorizationException: Not authorized to access topics: [t1]"
```